File: tools/t_class_classifier/python/t_signature.py
from schemas import RealizedClosureTrace, TSig
from feature_extract import compute_graph_metrics, estimate_braid_proxy

def build_t_sig(trace: RealizedClosureTrace) -> TSig:
    V, E, C, loop_count, raw_edge_count, parallel_edge_count = compute_graph_metrics(trace)
    
    # Calculate loop nesting depth. 
    if loop_count == 0:
        L_depth = 0
    elif loop_count == 1:
        L_depth = 1
    else:
        L_depth = 2

    # Connectivity persistence R_conn = E / V
    R_conn = float(E) / float(V) if V > 0 else 0.0

    # Braid index proxy
    B_cross = estimate_braid_proxy(trace, loop_count)

    return TSig(
        C_count=loop_count,
        L_depth=L_depth,
        R_conn=R_conn,
        B_cross=B_cross,
        component_count=C,
        raw_edge_count=raw_edge_count,
        unique_edge_count=E,
        parallel_edge_count=parallel_edge_count
    )

File: tools/t_class_classifier/cpp/t_signature.cpp
#include "schemas.hpp"

void compute_graph_metrics(
    const RealizedClosureTrace& trace,
    int& V,
    int& E,
    int& C,
    int& loop_count,
    int& raw_edge_count,
    int& parallel_edge_count
);
int estimate_braid_proxy(const RealizedClosureTrace& trace, int loop_count);

TSig build_t_sig(const RealizedClosureTrace& trace) {
    int V = 0, E = 0, C = 0, loop_count = 0, raw_edge_count = 0, parallel_edge_count = 0;
    compute_graph_metrics(trace, V, E, C, loop_count, raw_edge_count, parallel_edge_count);

    int L_depth = 0;
    if (loop_count == 0) {
        L_depth = 0;
    } else if (loop_count == 1) {
        L_depth = 1;
    } else {
        L_depth = 2;
    }

    double R_conn = 0.0;
    if (V > 0) {
        R_conn = static_cast<double>(E) / static_cast<double>(V);
    }

    int B_cross = estimate_braid_proxy(trace, loop_count);

    TSig sig;
    sig.C_count = loop_count;
    sig.L_depth = L_depth;
    sig.R_conn = R_conn;
    sig.B_cross = B_cross;
    sig.component_count = C;
    sig.raw_edge_count = raw_edge_count;
    sig.unique_edge_count = E;
    sig.parallel_edge_count = parallel_edge_count;
    return sig;
}
